pil-analyzer/src/traits_processor.rs

```diff
@@ -0,0 +1,139 @@
+        let trait_name = trait_impl.name.clone();
+
+        if types.len() != trait_decl.type_vars.len() {
+            panic!(
+                "{}",
+                trait_impl.source_ref.with_error(format!(
+                    "Trait {} has {} type parameters, but implementation has {}",
+                    trait_name,
+                    trait_decl.type_vars.len(),
+                    types.len(),
+                ))
+            );
+        }
+
+        let type_vars: HashSet<_> = trait_impl.type_scheme.vars.vars().collect();
+
+        let type_vars_in_tuple: Vec<_> = types
+            .iter_mut()
+            .flat_map(|t| {
+                t.map_to_type_vars(&type_vars);
+                t.contained_type_vars()
+            })
+            .collect();
+
+        let type_vars_in_scheme: Vec<_> = trait_impl.type_scheme.vars.vars().collect();
+
+        for var in type_vars_in_scheme {
+            if !type_vars_in_tuple.contains(&var) {
+                panic!(
+                    "{}",
+                    trait_impl.source_ref.with_error(format!(
+                        "Impl {trait_name} introduces a type variable {var} that is not used",
+                    ))
+                );
+            }
+        }
+    }
+}
+
+/// Ensures that there are no overlapping trait implementations in the given `implementations` map.
+///
+/// This function iterates through all the trait implementations comparing them with each other and ensure that
+/// there are no traits with overlapping type variables.
+fn ensure_unique_impls(implementations: &mut [TraitImplementation<Expression>]) {
+    for i in 0..implementations.len() {
+        let type_vars: HashSet<_> = implementations[i].type_scheme.vars.vars().collect();
+        implementations[i]
+            .type_scheme
+            .ty
+            .map_to_type_vars(&type_vars);
+
+        for j in (i + 1)..implementations.len() {
+            let type_vars: HashSet<_> = implementations[j].type_scheme.vars.vars().collect();
+            implementations[j]
+                .type_scheme
+                .ty
+                .map_to_type_vars(&type_vars);
+
+            unify_traits_types(
+                implementations[i].type_scheme.clone(),
+                implementations[j].type_scheme.clone(),
+            )
+            .map_err(|err| {
+                implementations[i]
+                    .source_ref
+                    .with_error(format!("Impls for {}: {err}", implementations[i].name))
+            })
+            .unwrap()
+        }
+    }
+}
+
+fn unify_traits_types(ty1: TypeScheme, ty2: TypeScheme) -> Result<(), String> {
+    let mut unifier = Unifier::new();
+    let instantiated_ty1 = unifier.instantiate_scheme(ty1).0;
+    let instantiated_ty2 = unifier.instantiate_scheme(ty2).0;
+
+    match unifier.unify_types(instantiated_ty1.clone(), instantiated_ty2.clone()) {
+        Ok(_) => Err(format!(
+            "Types {instantiated_ty1} and {instantiated_ty2} overlap"
```

Comment:
We should improve this error: First of all, we should use the non-instantiated type schemes (which use type variables familiar to the user), and then we should apply the substitutions to the type scheme to give an example type that applies to  both implementations.